NeXus–NOMAD metainfo generator: Phases 1 + 2 (#801)
FAIRmat-NFDI/pynxtools: docs/learn/pynxtools/nexus-metainfo-generation.md

Reviewer: There is no detector in the NXentry base def. https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/base_classes/NXentry.nxdl.xml

Author: Right, this was a mixup between `Entry` and `Object`. I changed it to a proper excerpt for now.

Reviewer: Are they really upper-cased, e.g., Class Xps(Entry)?

Author: That's old logic, they are title-cased (so `class Xps(Entry)` insted of `class XPS(Entry)`). I modified this part.

Reviewer: How it handles
Class A:
   name: # different desc and purpose
Class B(A):
   name:  # different desc and purpose 
Class C(B):
   name: # different desc and purpose

Author: It overwrites the name, just as in normal inheritance (and also how NeXus does it). The only collision it detects is if a field and a group (so a `Quantity` or a `Subsection`) would get the same name, which is not allowed in NOMAD.

- **Phase 1** (complete): 142 NXDL base classes (`category="base"`) generated and tested.
- **Phase 2** (complete): 85 application definitions (`category="application"`) generated; `nexus_applications` entry point active.

---

## Architecture

```
NexusNode API
    │  generate_tree_from(nx_name)
    │  → NexusDefinition root with NexusGroup/NexusField/NexusAttribute children
    ▼
nxdl_to_metainfo.py             ← generator
    │  build_context(nx_name)   ← builds Jinja2 template context
    │  render(context)          ← Jinja2 + ruff
    │  write_base_class(nx_name)
    ▼
metainfo/<folder>/<stem>.py ← generated Python files (one per NeXus definition) 
    ▼
metainfo/_package.py            ← assembles NOMAD Package
    │  build_package()
    ▼
NOMAD Package (section_definitions[])
```

---

## The NexusNode layer

### Why NexusNode, not raw XML

All NXDL attribute reading goes through the `NexusNode` API
(`pynxtools.nexus.nexus_tree`). No raw XML attribute access happens inside the
generator. This gives:

- **Inheritance resolution**: NexusNode follows the `extends` chain and merges attributes
  from parent classes. The generator never needs to know about inheritance.

> - **Stable surface**: if the NXDL XML format changes, only `nexus_tree.py` needs to

Reviewer 2: ```suggestion
- **Stable surface**: If the NXDL XML format changes, only `nexus_tree.py` needs to
```

| `ignore_extra_groups` | `bool` | `@ignoreExtraGroups` |
| `ignore_extra_fields` | `bool` | `@ignoreExtraFields` |
| `ignore_extra_attributes` | `bool` | `@ignoreExtraAttributes` |
| `symbols` | `dict[str, str]` | `<symbols>/<symbol>` elements |

#### `NexusGroup`

| Attribute | Type | Source (NXDL) |
|-----------|------|--------------|
| `nx_class` | `str` | `@type` |
| `occurrence_limits` | `tuple[int\None, int\None]` | `@minOccurs`/`@maxOccurs` |

#### `NexusField`

| Attribute | Type | Source (NXDL) |
|-----------|------|--------------|
| `dtype` | `str` | `@type` (default `"NX_CHAR"`) |
| `unit` | `str \ None` | `@units` |
| `shape` | `tuple \ None` | `<dimensions>` element |
| `dim_symbols` | `tuple \ None` | symbol names parallel to `shape` |
| `items` | `list[str] \ None` | `<enumeration>/<item>` |
| `open_enum` | `bool` | `<enumeration open="true">` |
| `long_name` | `str \ None` | `@long_name` |
| `interpretation` | `str \ None` | `@interpretation` |

#### `NexusAttribute`

Shares `dtype`, `shape`, `dim_symbols`, `items`, `open_enum` with `NexusField`.
Has no `unit`, `long_name`, `interpretation`, or plotting-hint attributes.

---

## Annotation models

> Six `AnnotationModel` subclasses (one per NXDL node kind) carry all NeXus-specific metadata on generated sections and quantities . Each annotation is queryable

Reviewer 2: ```suggestion
Six `AnnotationModel` subclasses (one per NXDL node kind) carry all NeXus-specific metadata on generated sections and quantities. Each annotation is queryable
```

at runtime via `m_def.m_get_annotations("nexus_field")` etc.

### `NeXusDefinition`

Attached to `Section.m_def` of every top-level generated class.

```python
class NeXusDefinition(AnnotationModel):
    nx_class: str
    category: Literal["base", "application", "contributed"] = "base"
    restricts: bool = False
    ignore_extra_groups: bool = False
    ignore_extra_fields: bool = False
    ignore_extra_attributes: bool = False
    symbols: dict[str, str] | None = None
    deprecated: str | None = None
```

### `NeXusGroup`

On named concept class `m_def` (group with own quantities, same file) or on the
`SubSection` for cross-file group references.

```python
class NeXusGroup(AnnotationModel):
    nx_class: str
    name: str | None = None
    name_type: Literal["specified", "any", "partial"] = "specified"
    optionality: Literal["required", "recommended", "optional"] = "optional"
    min_occurs: int | None = None
    max_occurs: int | None = None
    deprecated: str | None = None
```

### `NeXusField`

Attached to every `Quantity` derived from a NXDL `<field>` element. Fields

> carry typed data arrays and may have unit categories, interpretation hints,

Reviewer 2: single type

Author: I fixed this sentence to better show what fields are:
"""
Fields define the
type and shape of associated data, which may be scalar or array-valued. Fields may have
unit categories, interpretation hints, and long names.
"""

```

### `NeXusAttribute`

Attached to every `Quantity` derived from a NXDL `<attribute>` element (either
group-level or field-level). Field-level attributes set `parent_field` to the
owning field name (e.g. `parent_field="energy"` for `energy__units`).

```python
class NeXusAttribute(AnnotationModel):
    name: str
    parent_field: str | None = None
    type: str = "NX_CHAR"
    name_type: Literal["specified", "any", "partial"] = "specified"
    optionality: Literal["required", "recommended", "optional"] = "optional"
    enumeration: list[str] | None = None
    open_enum: bool = False
    deprecated: str | None = None
```

### `NeXusLink`

Attached to a `Quantity(type=str)` derived from a NXDL `<link>` element.
The `target` is the schema-level default target path; the parser resolves the
actual HDF5 target at read time.

```python
class NeXusLink(AnnotationModel):
    name: str
    target: str
    optionality: Literal["required", "recommended", "optional"] = "optional"
    deprecated: str | None = None
```

### `NeXusChoice`

Attached to each `SubSection` representing one alternative in a NXDL `<choice>`
block. A choice allows exactly one NX class to occupy a named slot. One

> SubSection is generated per alternative; all share the same `group_name`.

Reviewer 2: ```suggestion
`SubSection` is generated per alternative; all share the same `group_name`.
```

|---|---|---|
| `<definition>` | top-level Section class `m_def` | `nexus_definition` |
| `<group>` (named concept) | concept class `m_def` | `nexus_group` |
| `<group>` (cross-file) | `SubSection` | `nexus_group` |
| `<field>` | `Quantity` | `nexus_field` |
| `<attribute>` | `Quantity` | `nexus_attribute` |
| `<link>` | `Quantity(type=str)` | `nexus_link` |
| `<choice>` alternative | `SubSection` | `nexus_choice` |

---

## Generated file layout

```
src/pynxtools/nomad/
    annotations.py              # NeXusDefinition + NeXusGroup + NeXusField + NeXusAttribute
                                #   + NeXusLink + NeXusChoice + registration
    converters/
        __init__.py
        _mapping.py             # nxdl_to_class_name, BASESECTIONS_MAP, type mapping
        nxdl_to_metainfo.py     # generator: NexusNode → .py via Jinja2
        templates/
            nexus.py.j2    # Jinja2 template
    metainfo/
        __init__.py             # public API: build_base_classes_package(), build_applications_package()
        _package.py             # assembles NOMAD Packages
        base_classes/           # 142 generated base class files
            __init__.py

>             entry.py            # Entry(Object, basesections.Measurement)

Reviewer 2: not every entry is a measurement examples `NXmicrostructure_score_{config,results}`, `NXem`, `NXapm` with only `simulation` filled.

Author: Again, for now we let entry inherit from `basesections.Measurement`. to get it consistent across all classes I think in a follow-up PR we should probably go `Entry(Object, basesections.Activity)` and then each individual application definition either inherits from `basesections.Measurement` or, for simulation-focused classes, from `basesections.Actitvity` (at least until a `Simulation` class exists.

Reviewer 2: `basesections.Activity` I think is the best neutral enough choice, agreed

            sample.py           # Sample(Object, basesections.CompositeSystem)
            detector.py
            ...
        applications/           # 85 generated application definition files
            __init__.py
            arpes.py            # Arpes(Entry) — application specialization of Entry
            xps.py              # Xps(Entry) — application specialization of Entry
            ...
```

---

## Application Definitions

### The design principle: `Xps(Entry)` not `Xps(Object, Measurement)`

In NeXus, every application definition (e.g. `NXxps.nxdl.xml`) wraps exactly one `NXentry` at

> the root level. This is supposed to be a container pattern. However, in practice, no application definition has anything outside of its `NXentry` group. Therefore, application definitions can be understood as **constraint templates**

Reviewer 2: `NXroot` attribute are above `NXentry` motivating one additional annotation model instance ?

Author: These are all the same annotation model (`NexusDefinition`), but `NXroot` gets its own inheritance from `EntryData` (see #810) to group entries from one file into an overarching entry.

Reviewer 2: Yeah saw this later in the PR, accepted

> that describe what an `NXentry` looks like in a specific measurement context.

Reviewer 2: Already mentioned, here last time not every NXentry always a `measurement` can be a `simulation`.

Author: See above, this can be changed in a follow-up.

Therefore, in Python/NOMAD, application definitions inherit from `Entry` directly:

```python
class Xps(Entry):
    m_def = Section(
        a_nexus_definition=NeXusDefinition(nx_class="NXxps", category="application"),
    )
    # SubSections and Quantities from inside NXxps's NXentry group,
    # with XPS-specific optionality
```

> - `Xps` **is** a specialized kind of Entry, not a wrapper around an entry.

Reviewer 2: ```suggestion
- `Xps` **is** a specialized kind of `Entry`, not a wrapper around an entry.
```

- Parser logic is simpler: directly instantiate `Xps` when an HDF5 file's `NX_class` attribute says `"NXxps"`.

> - Inheritance chain is clear: `Object → Entry → Xps`.

Reviewer 2: Agreed

    NXdata.
    """

    m_def = Section(
        links=[
            "https://fairmat-nfdi.github.io/nexus_definitions/classes/base_classes/NXentry.html#nxentry-thumbnail-group"
        ],
        a_nexus_group=NeXusGroup(
            nx_class="NXnote",
            name="thumbnail",
            name_type="specified",
            optionality="optional",
        ),
    )

    type = Quantity(
        type=MEnum(["image/*"]),
        links=[
            "https://fairmat-nfdi.github.io/nexus_definitions/classes/base_classes/NXentry.html#nxentry-thumbnail-type-attribute"
        ],
        description=("The mime type should be an ``image/*``"),
        a_nexus_attribute=NeXusAttribute(
            name="type",
            type="NX_CHAR",
            name_type="specified",
            optionality="optional",
            enumeration=["image/*"],
            deprecated="Use the `type` field instead",
        ),
        a_eln=ELNAnnotation(
            component=ELNComponentEnum.EnumEditQuantity,
            default="image/*",
        ),
    )

    def normalize(self, archive: EntryArchive, logger: BoundLogger) -> None:
        super().normalize(archive, logger)
```

> ### Named concepts vs. cross-file SubSections

Reviewer 2: ```suggestion
### Named concepts vs. cross-file `SubSections`

A named concept class is generated only when the group defines **its own quantities** that
differ from the generic class: changed optionality, extra fields, or different

> type/units/enumeration. Groups that exist purely for occurrence context (e.g. a

Reviewer 2: ```suggestion
`type/units/enumeration`. Groups that exist purely for occurrence context (e.g. a
```

Author: No, this is just a list of actual words.

`backgroundBACKGROUND` group with no additional fields) point their `SubSection` directly
at the generic class; no named concept class is generated. This rule also prevents
circular imports since every class transitively extends `Object` (the generated `NXobject`
class).

| SubSection target | `a_nexus_group` location |
|---|---|
| Class defined in **this same file** (named concept) | On the concept class `m_def` |
| Class defined in **another file** (cross-file reference) | On the `SubSection` |

This means: named concept classes fully self-describe their occurrence context.

> Cross-file SubSections carry the occurrence context because the target class is generic

Reviewer 2: ```suggestion
Cross-file `SubSections` carry the occurrence context because the target class is generic
```

    ...
)
```

`NX_ANY` is different: it means the field's unit is **not fixed by the
schema at all** — the actual unit varies per file/value. There is no
`dimensionality`/`unit` to assign, so instead the generated `Quantity` gets
`flexible_unit=True` — a first-class NOMAD `Quantity` parameter for exactly
this case ("this quantity may have a unit that is not the default unit");
NOMAD then stores the value in full-storage mode (`MQuantity`), preserving
whichever unit was actually written:

```python
# NXDL: <field name="resolution" type="NX_FLOAT" units="NX_ANY">
resolution = Quantity(
    type=np.float64,
    flexible_unit=True,
    ...
    a_eln=ELNAnnotation(component=ELNComponentEnum.NumberEditQuantity),
)
```

No special ELN handling is needed beyond the usual `NumberEditQuantity`: its
GUI implementation already supports typing a value together with any unit
(and a unit-selection dropdown) when the field has no fixed dimensionality —
exactly the `NX_ANY` case.

---

## Additive-only generation

The generator never removes hand-written additions. When a file already exists:

- If `force=False` (default): the file is rewritten whenever the generated source differs

>   from the existing file **and** the existing file contains no user-added members (methods

Reviewer 2: So also not rewritten partially, right?

Author: It does get rewritten partially if e.g, the docstring in the NeXus definition changes. However, user-added concepts and `normalize` functions never get overwritten with `force=False`.

  or quantities not present in the new output). This means generator-driven changes
  (descriptions, annotation values, new members) propagate automatically, while files with
  custom `normalize()` logic are protected.
- If `force=True`: the file is always overwritten.
- In `dry_run` mode: returns `True` if the file would differ; no disk write occurs.

> User-added members are detected by comparing the set of names in the existing file against

Reviewer 2: Example would help here

Author: I am adding user-added members in #810 in `entry-py` and `root.py`. I will add an example there.

the set the generator would produce. Any name present in the existing file but absent from
the new output is considered user-added.

---

## CLI

```bash
# Generate one class (base or application)
pynx nomad generate-metainfo --nxdl NXentry
pynx nomad generate-metainfo --nxdl NXarpes

# Generate all base classes
pynx nomad generate-metainfo --all-base

# Generate all application definitions
pynx nomad generate-metainfo --all-applications

# Generate all categories (apps first, then base --force for cross-refs)
pynx nomad generate-metainfo --all

# CI check: fail if committed files differ from what the generator would produce
pynx nomad generate-metainfo --all --dry-run

# Force regeneration (overwrite existing files)
pynx nomad generate-metainfo --all --force

# Generate into a different package (e.g. nomad-measurements)
pynx nomad generate-metainfo --all --output-dir ../nomad-measurements/nexus/metainfo/base_classes
```

---

## Package assembly

`metainfo/_package.py` imports all generated modules and assembles NOMAD `Package` objects:

### Base classes package

```python
from pynxtools.nomad.metainfo import build_base_classes_package

pkg_base = build_base_classes_package()

> print(len(pkg_base.section_definitions), "base sections")  # 327 including named concepts

Reviewer 2: Would remove the counts as these are subject to change unless automatically updated.

Author: Makes sense, counts removed everywhere

```

### Applications package

```python
from pynxtools.nomad.metainfo import build_applications_package

pkg_apps = build_applications_package()

> print(len(pkg_apps.section_definitions), "application sections")  # 345 including named concepts

Reviewer 2: also here and in one place at the top of this file

Author: Makes sense, counts removed everywhere

```

During package assembly, `__init_metainfo__()` resolves all `SubSection.section_def`

> string FQNs into live class references.

Reviewer 2: ```suggestion
string fully qualified names (FQNs) into live class references.
```

Reviewer 2: ?

Author: yes thanks

---

## Inheritance chain

Every generated class extends its **NeXus parent** class (the Python class generated from
the NXDL `extends` attribute) and — where a semantic match exists — additionally inherits

> from the corresponding NOMAD base section:

Reviewer 2: ```suggestion
from the corresponding NOMAD `BaseSection`
```

?

Author: No, `basesections.BaseSection` is a specific class, what is meant here is "one of the base sections", i.e., as prose

```python
class Object(BaseSection): ...            # NXobject → root

> class Entry(Object, basesections.Measurement): ...

Reviewer 2: Already discussed

But I also see the issue in that https://fairmat-nfdi.github.io/nomad-simulations/schema/model_method/ is geared towards 
https://fairmat-nfdi.github.io/nomad-simulations/schema/simulation/

Author: I think the cleanest would probably to go ´ Entry(Object, basesections.Measurement)` and then for each application definition individually we decide if we want it to be a `basesections.Activity` or, more specified, a `basesections.Measurement`.

Note that not using `basesections.Measurement`.comes with a real cost though: the base section contains subsections for instruments, sample, and data + normalization on top of these. That means that any application definition that ineherits from `basesections.Activity` instead of `basesections.Measurement` **does not get these inherited terms and normalizers**.

To be discussed in a separate PR, this is too big for this one here.

Reviewer 2: Agree that this is a discussion where for this PR one should have the choices made elsewhere.
In my opinion in the long run it would be best if NOMAD provides a `basesection.Activity`, which gets specialized into a `basesection.Simulation` and a `basesection.Measurement`.
For simulations made by FAIRmat v1 Area D and all other folks doing simulations, they can use the same basesection to build from.  Simulations do not need an `instrument` section but rather `computer` as that is the `instrument`, and this is direction which Area D layed out. Agree though cross-area semantic harmonization.
Note for completion also that it is also not always straightforward to clearly separate the entry is a only measurement or simulation, often the NXprocess wraps what is effectively a simulation, typical case, electron diffraction, indexing a pattern where one combines a measured pattern with a simulated pattern. Not necessarily the simulation and measurement are stored as separate entries, so what is then such a Measurement which has Simulation parts. Despite this ambivalence I still convinced though it is useful if a system like NOMAD offers two clean suggestions what to do in case when I have primarily a measurement and or computer simulation.

class Sample(Component, basesections.CompositeSystem): ...
class SampleComponent(Component, basesections.Component): ...
...
```

The NOMAD base section is imported **as the module** (`from nomad.datamodel.metainfo import
basesections`) and used as `basesections.Measurement` in the class signature. This avoids
name collisions when the generated NeXus class and the NOMAD base share the same Python
name (e.g. both `Component`).

The secondary NOMAD base is only added when the NeXus `extends` chain does not already

> provide it;s preventing duplicate entries in the MRO.

Reviewer 2: ```suggestion
provide it; thus preventing duplicate entries in the method resolution order (MRO).
```

Reviewer 2: Spell out "method resolution order"

Author: Okay, I updated your suggestion to spell it out

| NXDL class | Generated class signature |
|-----------|--------------------------|
| `NXobject` | `Object(BaseSection)` |

> | `NXentry` | `Entry(Object, basesections.Measurement)` |

Reviewer 2: `basesections.Activity` ?

Author: We discussed this already, also above. I really don't think that we just be less specific than measurement here, especially if we move the schema to `nomad-measurements` soon. In any case, this automated map will be removed eventually.

(when absent) the NeXus class stem in uppercase (e.g. `"USER"` for any `NXuser`).
The three `name_type` values differ only in *casing*, not in which name is
used:

- `name_type="specified"` (a fixed-name group, e.g. `instrument`): used as-is
  — NXDL convention already writes these in snake_case.
- `name_type="any"` (a variadic group; instances may use any actual HDF5
  name): `node.name` is lowercased. This covers both an anonymous
  group with no `name=` attribute at all (e.g. any `NXdetector` → `detector`)
  *and* an explicitly-named variadic group (e.g. `name="BIAS_SWEEP"` →
  `bias_sweep`).
- `name_type="partial"` (e.g. `peakPEAK`): used as-is, preserving the mixed
  case, so the lower-case fixed prefix and the upper-case user-chosen portion
  both stay visible in the generated code (consistent with the matching
  concept class name, e.g. `FitPeakPEAK`).

The `a_nexus_group(name=...)` annotation, separately, faithfully records what
NXDL actually declares: `None` only for a genuinely anonymous group (no
`name=` attribute at all), the literal NXDL name otherwise — so `name=None`
for any `NXuser`, but `name="BIAS_SWEEP"` for the explicitly-named variadic

> group above, even though both get lowercased the same way for the Python

Reviewer 2: minor, lower-cased, lowercased

Author: changed

attribute name.

**One exception to "groups always win the unqualified name"**: a group

> literally named `name`, `datetime`, `lab_id`, or `description` gets a

Reviewer 2: I thought `lab_id` was replaced by `id` at some point, was this idea dropped ?

Author: This was never implemented and I don't think it will be replaced fully because many users rely on `lab_id` already. The only thing we can do is provide an alias called `id` in the future.

## Overview

The NeXus–NOMAD Metainfo refactor replaces the runtime XML compilation in `nomad/schema_packages/schema.py`
with a **persistent code generator** that reads the NeXus definitions through the `NexusNode` API
and writes one Python file per NXDL class.

### What this provides

- **Importable, type-checkable Python classes**: Each NXDL class becomes a NOMAD `Section` class
  in a `.py` file, with structured annotations connecting every quantity and group back to NXDL.
- **Clean naming**: Quantities are named after their NXDL concept (`start_time`, not `start_time__field`).
- **Two entry points**: `nexus_base_classes` (142 base classes) and `nexus_applications` (85 application definitions).
- **Application definitions are Entry subclasses**: `Xps(Entry)` instead of wrappers — see [ADR-006](#adr-006-application-definitions).

> - **Backward compatibility**: The old `nexus_schema` entry point is untouched; both systems coexist until Phase 3.

Reviewer: Does it mean in-memory nexus_schema genereated by nomad/schema_pachages/(schema.py and converter.py) will be available?

Author: For now, yes. The idea is to have the new schemas/parsers in parallel for a while, until we are sure the new versions are working properly. Only then we remove the old `schema.py`/`parser.py`

            detector.py
            ...
        applications/           # Phase 2: 85 generated application definition files
            __init__.py
            arpes.py            # Arpes(Entry) — application specialization of Entry
            xps.py              # Xps(Entry) — application specialization of Entry
            ...
```

---

## Application Definitions: The XpsEntry Decision {#adr-006-application-definitions}

### The design principle: `Xps(Entry)` not `Xps(Object, Measurement)`

In NeXus, every application definition (e.g. `NXxps.nxdl.xml`) wraps exactly one `NXentry` at
the root level. This is not a container pattern — application definitions are **constraint templates**
that describe what an `NXentry` looks like in a specific measurement context.

Therefore, in Python/NOMAD, application definitions inherit from `Entry` directly:

```python
# Phase 2: generated application definition
class Xps(Entry):
    m_def = Section(
        a_nexus_definition=NeXusDefinition(nx_class="NXxps", category="application"),
    )
    # SubSections and Quantities from inside NXxps's NXentry group,
    # with XPS-specific optionality
```

> **Why this design** (see [ADR-006](../../data-modeling/nexus-metainfo/adr/ADR-006-application-definitions-are-entry-subclasses.md)):

Reviewer: These links are coming from your local folder structure.

Author: I just removed them.
